dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,215 @@
+ */
+template <typename Out, typename In, typename W, int ndim, int axis, bool has_channels = true>
+struct ConvolutionGpu {
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const TensorListShape<1>& window_size) {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string("Provided input shape and window sizes do not mach in number of samples: ",
+                    in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(window_size[i][0] % 2 == 1,
+                   make_string("Kernel window should have odd length, got: ", window_size,
+                               " for sample ", i, "."));
+
+      DALI_ENFORCE((window_size[i][0] / 2) * num_channels < kMaxRadiusSpan,
+                   make_string("Kernel window too big for sample ", i, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& window, W scale = 1) {
+    int num_samples = in.size();
+    auto* window_tmp_buffer_host =
+        ctx.scratchpad->Allocate<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    auto* window_tmp_buffer_gpu =
+        ctx.scratchpad->Allocate<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+
+    // Pad and align windows in tmp memory, transfer the aligned windows to GPU
+    FillAlignedWindow(window_tmp_buffer_host, window, in.shape);
+    cudaMemcpyAsync(window_tmp_buffer_gpu, window_tmp_buffer_host,
+                    sizeof(W) * num_samples * kWindowCopyBufferSize, cudaMemcpyHostToDevice,
+                    ctx.gpu.stream);
+
+    Arguments args;
+    if (kIsInnerConv) {
+      // Inner (innermost) - repack arguments
+      for (int i = 0; i < num_samples; i++) {
+        cutlass::Array<int, 2> size;
+        auto sample_shape = in.tensor_shape(i);
+        int num_channels = has_channels ? sample_shape[ndim - 1] : 1;
+        // height
+        size[0] = volume(sample_shape.begin(), sample_shape.begin() + axis);
+        // width
+        size[1] = sample_shape[ndim - has_channels - 1];
+        int row_stride = sample_shape[ndim - has_channels - 1] * num_channels;
+        auto* window_gpu =
+            reinterpret_cast<cutlass_W*>(window_tmp_buffer_gpu + i * kWindowCopyBufferSize);
+        int window_diameter = static_cast<int>(window.tensor_shape_span(i)[0]);
```

Comment:
```suggestion
        int window_size = static_cast<int>(window.tensor_shape_span(i)[0]);
```
Even if we had a "radius", goiing to diameter is a stretch.

Reply:
done